pypit/wavetilts.py

```diff
@@ -0,0 +1,480 @@
+
+        """
+        # Arc
+        msarc_file = armasters.core_master_name('arc', setup, mdir)
+        msarc, _, _ = armasters._load(msarc_file)
+
+        # Instantiate
+        slf = cls(msarc)
+
+        # Tilts
+        mstilts_file = armasters.core_master_name('tilts', setup, mdir)
+        hdul = fits.open(mstilts_file)
+        slf.final_tilts = hdul[0].data
+        slf.tilts = slf.final_tilts
+
+        # Dict
+        slf.all_trcdict = []
+        islit = 0
+        for hdu in hdul[1:]:
+            if hdu.name == 'FWM{:03d}'.format(islit):
+                # Setup
+                fwm_img = hdu.data
+                narc = fwm_img.shape[1]
+                trcdict = dict(xtfit=[], ytfit=[], xmodel=[], ymodel=[], ycen=[], aduse=np.zeros(narc, dtype=bool))
+                # Fill  (the -1 are for ycen which is packed in at the end)
+                for iarc in range(narc):
+                    trcdict['xtfit'].append(fwm_img[:-1,iarc,0])
+                    trcdict['ytfit'].append(fwm_img[:-1,iarc,1])
+                    trcdict['ycen'].append(fwm_img[-1,iarc,1])  # Many of these are junk
+                    if np.any(fwm_img[:-1,iarc,2] > 0):
+                        trcdict['xmodel'].append(fwm_img[:-1,iarc,2])
+                        trcdict['ymodel'].append(fwm_img[:-1,iarc,3])
+                        trcdict['aduse'][iarc] = True
+                #
+                slf.all_trcdict.append(trcdict.copy())
+            else:
+                slf.all_trcdict.append(None)
+            islit += 1
+        # FInish
+        return slf
+
+
+    def _analyze_lines(self, slit):
+        """
+        Analyze the tilts of the arc lines in a given slit/order
+
+        Wrapper to artracewave.analyze_lines()
+
+        Parameters
+        ----------
+        slit : int
+
+        Returns
+        -------
+        self.badlines
+
+        """
+        self.badlines, self.all_ttilts[slit] = artracewave.analyze_lines(
+            self.msarc, self.all_trcdict[slit], slit, self.pixcen, self.settings)
+        if self.badlines > 0:
+            msgs.warn("There were {0:d} additional arc lines that should have been traced".format(self.badlines) +
+                      msgs.newline() + "(perhaps lines were saturated?). Check the spectral tilt solution")
+        # Step
+        self.steps.append(inspect.stack()[0][3])
+        return self.badlines
+
+    def _extract_arcs(self):
+        """
+        Extract the arcs down each slit/order
+
+        Wrapper to ararc.get_censpec()
+
+        Returns
+        -------
+        self.arccen
+        self.arc_maskslit
+
+        """
+        # Extract an arc down each slit/order
+        self.arccen, self.arc_maskslit, _ = ararc.get_censpec(self.lordloc, self.rordloc,
+                                                              self.pixlocn, self.msarc, self.det,
+                                                              gen_satmask=False)
+        self.satmask = np.zeros_like(self.msarc)
```

Comment:
This seems a little bit too deep. Should we have gen_satmask as a variable given to the class?

Reply:
Now accessible with run()